ODIN_II/SRC/include/MixingOptimization.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#ifndef MIXING_OPTIMIZATION_HPP
+#define MIXING_OPTIMIZATION_HPP
+#include "odin_types.h" // netlist_t, config_t
+/**
+ * @brief A base class in hierarchy for complex synthesis
+ * allowing for mixing soft and hard logic 
+ */
+class MixingOpt {
+  public:
+    /**
+     * @brief Construct a new Mixing Opt object
+     * 
+     * By default, all optimizations only share 
+     * the ratio of blocks to be implemented in 
+     * hard logic
+     * @param ratio, a value within 0 to 1 to 
+     * implement ratio*requested hard blocks in
+     * hard logic
+     * @param kind a kind of blocks that correspond
+     * to optimization pass
+     */
+    MixingOpt(float ratio, operation_list kind)
+        : _kind(kind) {
+        _ratio = ratio;
+        _enabled = true;
+    };
+    /**
+     * @brief Destroy the Mixing Opt object
+     * required by compiler
+     */
+    virtual ~MixingOpt() = default;
+
+    /**
+     * @brief performs the optimization pass, varies between kinds.
+     * If the implementation is not provided within the inherited class
+     * will throw ODIN error
+     * 
+     * @param netlist_t* pointer to a global netlist
+     * @param std::vector<nnode_t*> a vector with nodes the optimization
+     * pass is concerned (all of which are potential candidates to 
+     * be implemented in hard blocks for a given _kind)
+     */
+    virtual void perform(netlist_t*, std::vector<nnode_t*>&);
+    /**
+     * @brief Set the blocks of blocks required
+     * by counting in netlist
+     *
+     * @param count 
+     */
+    virtual void set_blocks_needed(int count);
+
+    bool enabled() {
```

Comment:
add an argument ` nnode_t * /*node*/` it will be unused in most case, but some will need it

Reply:
Done, thanks for the comment!

Reply:
Done, thanks for the comment!